### PR TITLE
[word lo/hi] docs and more word utilities

### PR DESCRIPTION
**Pull request by hero78119**
### Description

Follow https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1435 design duplicate field design.

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Contents
- more utilities on `Word` type, and further refactor few to be more generics

### CODE REVIEW

File: zkevm-circuits/src/util/word.rs [hunk 1]
```diff
@@ -100,6 +117,90 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
...
+                bytes_lo_assignedcells.to_vec(),
+                bytes_hi_assignedcells.to_vec(),
+            ]
+            .concat()),
+            _ => Err(Error::Synthesis),
+        }
+    }
+
+    /// assign u256 to wordlimbs
+    pub fn assign_u256(
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/util/word.rs:172]: Does this function work for only ` WordLimbs<Cell<F>, 2>` or does it work for  `WordLimbs<Cell<F>, 4>`, `WordLimbs<Cell<F>, 8>`, `WordLimbs<Cell<F>, 16>`, and `WordLimbs<Cell<F>, 32>`.

We only have at most 5 cases, but how we do it seems very involving.

File: zkevm-circuits/src/util/word.rs [hunk 2]
```diff
@@ -100,6 +117,90 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
...
+        offset: usize,
+        bytes_lo: Option<[u8; N_LO]>,
+        bytes_hi: Option<[u8; N_HI]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert_eq!(N % 2, 0); // TODO use static_assertion instead
+        assert_eq!(N_LO % (N / 2), 0);
+        assert_eq!(N_HI % (N / 2), 0);
+        // assign lo
+        let bytes_lo_assigned: Result<Vec<AssignedCell<F, F>>, Error> =
+            bytes_lo.map_or(Err(Error::Synthesis), |bytes| {
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/util/word.rs:141]: We can use the question mark `?` to map the error out.

**Review by ChihChengLiang — CHANGES_REQUESTED** [review]
I added some comments for the overall designs.
Those designs would have a big impact on the whole refactor, so I feel the need to have more discussion.

File: zkevm-circuits/src/witness/rw.rs [hunk 3]
```diff
@@ -246,35 +247,49 @@ pub enum Rw {
 
 /// Rw table row assignment
 #[derive(Default, Clone, Copy, Debug)]
-pub struct RwRow<F> {
+pub struct RwRow<F, T> {
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs:250]: Intuitively, we only need `F`, and `Word<F>`. So there is no point to introduce `T` here. 

However, Introduce this generics is a workaround, to handle this `https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/rw.rs#L421` 
Which we have `RwRow<Value<F>>` here. To support `RwRow<Value<F>, Value<Word<F>>>` a workaround is to generic type `T`. If have better idea I will be happy to remove this generics.

File: zkevm-circuits/src/witness/rw.rs [hunk 4]
```diff
@@ -246,35 +247,49 @@ pub enum Rw {
...
-            self.aux2,
+            vec![
+                self.rw_counter,
+                self.is_write,
+                self.tag,
+                self.id,
+                self.address,
+                self.field_tag,
+            ],
+            self.storage_key_word.clone().to_vec(),
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs:284]: Because `RwRow`  can't introduce `Word<F>` that's why here need to use vector way. Also see explanation here https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/files#r1211032980

File: zkevm-circuits/src/util/word.rs [hunk 5]
```diff
@@ -231,6 +364,11 @@ impl<F: Field> Word<Expression<F>> {
     pub fn sub_unchecked(self, rhs: Self) -> Self {
         Word::new([self.lo() - rhs.lo(), self.hi() - rhs.hi()])
     }
+
+    /// No overflow check
+    pub fn expr_unchecked(&self) -> Expression<F> {
+        self.lo() + self.hi() * (1 << (N_BYTES_HALF_WORD * 8)).expr()
+    }
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs:371]: There are many, can see one example in evm circuit word lo/hi refactor [TransferGadget](`https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1411/files#diff-e72a6a818cbda6c14cb551573432e59fe61ce109345dd4c088508a60c27979afL178-L183`) Just be patient 😄 It took 10 secs on my browser to nativage the the cursor haha

Basically this is to handle `address`, such as caller_address, callee_address, coinbase. We have many gadgets take address as single expression. This function is to use to convert address from word to single expression. And who use/review this function to be aware.  

Our principle is, if there is `Word<Expression<F>>` type and need to put address, then it should be 16 bytes on `lo` and 4 bytes on `hi`.  But if there just require a single expression, we can convert address to single expression with this function.

File: zkevm-circuits/src/witness/rw.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply zkevm-circuits/src/witness/rw.rs:250]: The `RwRow<Value<F>>` seems fixable. `Word<F>` is also compatible with `Word<Value<F>>`. I tried the idea, and it builds. Check this out.

https://github.com/privacy-scaling-explorations/zkevm-circuits/compare/feat/word-lo-hi-docs-util...review-commments-word-lo-hi-docs-util?expand=1

File: zkevm-circuits/src/witness/rw.rs [hunk 6]
```diff
@@ -64,19 +64,20 @@ impl RwMap {
...
                 if is_first {
                     // value == init_value
-                    let init_value = updates
+                    let _init_value = updates
                         .get(row)
                         .map(|u| u.value_assignments(mock_rand).1)
                         .unwrap_or_default();
-                    if value != init_value {
-                        errs.push((idx, err_msg_first, *row, *prev_row));
+                    todo!("use word equal value == init_value instead");
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs:74]: It's a good call! thanks and will adopt it

File: zkevm-circuits/src/witness/rw.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs]: Thanks for the kindly suggest, I applied your method and rollback unnessary modifcation 👍

File: zkevm-circuits/src/witness/rw.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs]: Fixed in commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/51a89bb3939e1c148515cf7a5badcf173cc2f565

File: zkevm-circuits/src/witness/rw.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs]: Fixed this issue mention in comment :) https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441#discussion_r1211062212

File: zkevm-circuits/src/util/word.rs (hunk 2 above)
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs]: Rewrited this part to partially adopt the proposed way https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/d973f4aa9faee167179a60c6115c97b44d3944c4

File: zkevm-circuits/src/util/word.rs [hunk 7]
```diff
@@ -100,6 +117,90 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
...
+    // N_HI = 2 => [nh1, nh2]
+    // N = 2 => [l1, l2]
+    // it equivalent l1.assign(nl1.expr() + nl2.expr() * 256 + nl3.expr() * 256^2 +  nl3.expr() *
+    // 256^3) and l2.assign(nh1.expr() + nh2.expr() * 256)
+    pub fn assign_lo_hi<const N_LO: usize, const N_HI: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes_lo: Option<[u8; N_LO]>,
+        bytes_hi: Option<[u8; N_HI]>,
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs]: My option is assign lo/hi separately still better than padding way since it do not need to waste extra space just for zero padding. Besides it pretty flexible. 

It just used by `assign_u256`, `assign_h160`, `assign_u64`, so I changed visibility to `private` in case of confusing.

File: zkevm-circuits/src/util/word.rs (hunk 1 above)
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs:168]: it works for all `WordLimbs<Cell<F>, N>` and inner function `assign_lo_hi` will help to handle all the case

File: zkevm-circuits/src/util/word.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply zkevm-circuits/src/util/word.rs:373]: Thanks for pointing out the link.
I checked out the use cases and now see the rationale for having this function.

However, I still think the use `expr_unchecked` doesn't make sense most of the time, or at least we shouldn't use them directly in the circuit. There are some reasons:

With the word lo-hi refactor, we can represent addresses as words in as many places as possible to minimize the effort for conversion. It would be something similar In spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/424/.

One of the reasons we need to use `expr_unchecked` to convert an address from a word is that many of our API returns address as a word. For example, we got `callee_address` as a word because the API `call_context_read_as_word` now returns a word.

As address as a word output become common, it would be convenient for the API to take an address as a word as input. The API knows that the input should be an address, so doing the  `expr_unchecked` conversion inside that API is safe. Pushing the conversion decision to circuit developers is more error-prone.

I checked the use cases in the evm circuit PR, the conversions happen in gadgets like UpdateBalanceGadget and TransferGadget and APIs like `account_storage_read` or `account_read`. They all propagate the address all the way to RWtable lookup. So one best destination for the conversion might be at the rw_lookup API.

If we look at the [RWTable](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#rw_table), the only place we can consider compressing the address from a word to a field element is "4 Address k2", which is this place https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/review-commments-word-lo-hi-docs-util/zkevm-circuits/src/evm_circuit/table.rs#L332 in the circuit.

To summarize, I propose the following:

- we change the function signatures of whatever API takes addresses as inputs to take them as a word.
- we compress the address at the level of  `fn input_exprs(&self)` of the `Lookup` struct or somewhere that low level.

File: zkevm-circuits/src/util/word.rs [hunk 8]
```diff
@@ -75,6 +78,12 @@ impl<F: Field> Expr<F> for Word32Cell<F> {
     }
 }
 
+/// Get word from Lo
+pub trait FromLo<T: Default> {
+    /// from lo generic type T unchecked
+    fn from_lo_unchecked(lo: T) -> Word<T>;
+}
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/util/word.rs:85]: What is the intended use case for this trait?

This trait is more useful if we can implement it for `Field` and `Value<Field>`. But using that trait for  `Field` and `Value<Field>` to implicitly convert unchecked field numbers seems insecure.

I propose we should have a constructor in `Word` to construct the word from lo limb.

File: zkevm-circuits/src/witness/rw.rs [hunk 9]
```diff
@@ -561,69 +574,43 @@ impl Rw {
...
-            }),
-            Self::AccountStorage { value_prev, .. } => {
-                Some(rlc::value(&value_prev.to_le_bytes(), randomness))
-            }
+            Self::Account { value_prev, .. } => Some(WordNew::from_u256(*value_prev)),
+            Self::AccountStorage { value_prev, .. } => Some(WordNew::from_u256(*value_prev)),
             Self::TxAccessListAccount { is_warm_prev, .. }
             | Self::TxAccessListAccountStorage { is_warm_prev, .. } => {
-                Some(F::from(*is_warm_prev as u64))
+                Some(WordNew::new([F::from(*is_warm_prev as u64), F::ZERO]))
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/witness/rw.rs:609]: Can the from_u64 constructor work here?
```suggestion
                Some(WordNew::from_u64(*is_warm_prev as u64))
```

File: zkevm-circuits/src/witness/rw.rs [hunk 10]
```diff
@@ -561,69 +574,43 @@ impl Rw {
...
+            }
+            Self::TxRefund { value_prev, .. } => {
+                Some(WordNew::new([F::from(*value_prev), F::ZERO]))
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/witness/rw.rs:612]: Same here.
```suggestion
                Some(WordNew::from_u64(*value_prev))
```

**Review by ChihChengLiang — COMMENTED** [review]
Added comments on constructors and traits

File: zkevm-circuits/src/util/word.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply zkevm-circuits/src/util/word.rs]: Sounds good to me.

**Comment by rrtoledo** [comment]
Made a pass, found a typo and and mention ways to make the code a bit more concise.

File: zkevm-circuits/src/util/word.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs:373]: Pretty agree with your proposal! For split address in word hi/lo, why we need to prioritize it is because it's more like a bug since in word expression we do not expect to have lo > 128 bits. Here is more like a consistent address representation across the codebase.

How about 
- mark this function as `deprecated` for the unsafe reason for now.
- then probably removed it in the following evm_circuit pr along with rw table fix ~?

File: zkevm-circuits/src/util/word.rs (hunk 8 above)
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs]: Right, originally I didn't put into constructors is  because it will conflict the `Word<Expression<F>> ` for `Expression` do not derive `Default` trait. But finally i figure out even define this trait will still conflict with `Expression`. 

Remove this trait entirely!

File: zkevm-circuits/src/witness/rw.rs (hunk 9 above)
Comment (hero78119) [reply zkevm-circuits/src/witness/rw.rs]: fixed in latest commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/f4cc21fb96ebece18e5b9cbf656ee63b94a3c609

File: zkevm-circuits/src/util/word.rs [hunk 11]
```diff
@@ -100,6 +109,86 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
...
+    // 256^3) and l2.assign(nh1.expr() + nh2.expr() * 256)
+    pub fn assign_lo_hi<const N_LO: usize, const N_HI: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes_lo: Option<[u8; N_LO]>,
+        bytes_hi: Option<[u8; N_HI]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert_eq!(N % 2, 0); // TODO use static_assertion instead
+        assert_eq!(N_LO % (N / 2), 0);
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs:130]: Keep N/2 in assertion part since later will inspect static assertion, and I believe static assertion not work on variable.

Adopt your suggestion after this assertion to make code looks nicer though

File: zkevm-circuits/src/util/word.rs [hunk 12]
```diff
@@ -231,6 +364,11 @@ impl<F: Field> Word<Expression<F>> {
     pub fn sub_unchecked(self, rhs: Self) -> Self {
         Word::new([self.lo() - rhs.lo(), self.hi() - rhs.hi()])
     }
+
+    /// No overflow check
+    pub fn expr_unchecked(&self) -> Expression<F> {
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/util/word.rs:369]:
```suggestion
    /// Compress the lo and hi limbs into an expression without checking the overflow.
    /// Only use it for address. We should remove it before merging to the main branch.
    #[deprecated]
    pub fn expr_unchecked(&self) -> Expression<F> {
```

File: zkevm-circuits/src/util/word.rs [hunk 13]
```diff
@@ -100,6 +109,86 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
...
+                bytes
+                    .chunks(N_LO / (N / 2)) // chunk in little endian
+                    .map(|chunk| from_bytes::value(chunk))
+                    .zip(self.limbs[0..(N / 2)].iter())
+                    .map(|(value, cell)| cell.assign(region, offset, Value::known(value)))
+                    .collect()
+            });
+
+        // assign hi
+        let bytes_hi_assigned: Result<Vec<AssignedCell<F, F>>, Error> =
```
[the remark this answers is not in the document]
Comment (rrtoledo) [reply zkevm-circuits/src/util/word.rs]: Sure :)

**Review by ChihChengLiang — APPROVED** [review]
LGTM. Great work adding the utilities, fixing the tables, and addressing all the feedback.

File: zkevm-circuits/src/util/word.rs (hunk 12 above)
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs]: cosmetics comment a bit and add deprecated msg in commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/84f96c38881e5cc4000c166e314242be80639c11 👌

**Comment by hero78119** [comment]
Hi @adria0 I updated this pr to rollback the changes in `zkevm-circuits/src/witness/rw.rs` and `zkevm-circuits/src/table.rs`, as its already covered by state circuit word lo-hi refactor pr https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1423

Would you kindly help to be second reviewer of this pr, thank you!

**Review by rrtoledo — APPROVED** [review]
LGTM, good work with the changes and thanks for addressing the reviews!

File: zkevm-circuits/src/util/word.rs [hunk 14]
```diff
@@ -136,11 +235,20 @@ impl<T: Clone> Word<T> {
...
     pub fn to_lo_hi(&self) -> (T, T) {
         (self.0.limbs[0].clone(), self.0.limbs[1].clone())
     }
+    /// Map the word to other types
+    pub fn map<T2: Clone>(&self, mut func: impl FnMut(T) -> T2) -> Word<T2> {
+        Word(WordLimbs::<T2, 2>::new([func(self.lo()), func(self.hi())]))
+    }
+}
+
+impl<T: Default> Default for Word<T> {
```
Comment (ed255) [inline zkevm-circuits/src/util/word.rs:248]: Isn't this equivalent to deriving `Default`?

File: zkevm-circuits/src/util/word.rs [hunk 15]
```diff
@@ -151,20 +259,46 @@ impl<T> std::ops::Deref for Word<T> {
...
 
+impl<T: Clone + PartialEq> PartialEq for Word<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.lo() == other.lo() && self.hi() == other.hi()
+    }
+}
+
 impl<F: Field> Word<F> {
-    /// Constrct the word from u256
+    /// Construct the word from u256
```
Comment (ed255) [inline zkevm-circuits/src/util/word.rs:269]: Maybe we could use the `From` trait for all these constructors?

**Review by ed255 — APPROVED** [review]
LGTM!  I've left some small comments, but feel free to merge anytime :)

File: zkevm-circuits/src/util/word.rs (hunk 15 above)
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs:269]: nice call and thanks!, i have addressed it in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/79025d72da404f0b9b0d74a0ab3f7b97881acc7c

File: zkevm-circuits/src/util/word.rs [hunk 16]
```diff
@@ -245,11 +245,11 @@ impl<T: Clone> Word<T> {
     }
 }
 
-impl<T: Default> Default for Word<T> {
-    fn default() -> Self {
-        Self(Word2::<T>::default())
-    }
-}
+// impl<T: Default> Default for Word<T> {
```
[the remark this answers is not in the document]
Comment (hero78119) [reply zkevm-circuits/src/util/word.rs]: y thanks, updated https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1441/commits/d88af85c028d7f692b749b78638154458a3c666b
